File: apps/android/core/analytics/src/main/kotlin/app/aistylist/core/analytics/ConsentGatedAnalytics.kt
package app.aistylist.core.analytics

import java.util.concurrent.atomic.AtomicBoolean

/**
 * Consent gate: OFF by default; nothing reaches [sink] until `setConsent(true)`, and nothing
 * tracked before opt-in is buffered or replayed.
 */
class ConsentGatedAnalytics(
    private val sink: AnalyticsSink,
) : Analytics {
    private val consent = AtomicBoolean(false)

    override fun hasConsent(): Boolean = consent.get()

    override fun setConsent(granted: Boolean) {
        consent.set(granted)
    }

    override fun track(event: AnalyticsEvent) {
        if (consent.get()) {
            sink.send(event)
        }
    }
}

/**
 * P02 sink: drops every event silently (no log, no network).
 * TODO(P03): PostHog Android adapter behind this same port; do not add the SDK before then.
 */
object NoopAnalyticsSink : AnalyticsSink {
    override fun send(event: AnalyticsEvent) = Unit
}

File: apps/ios/Packages/Core/Sources/Analytics/ConsentGatedAnalytics.swift
/// Current sink: drops every event silently (no logging, no network).
/// TODO(P03): PostHog iOS SDK adapter behind `AnalyticsSink`; do not add the dependency before.
public struct NoopSink: AnalyticsSink {
  public init() {}

  public func send(_ event: AnalyticsEvent) {}
}

/// Consent gate: OFF by default; nothing reaches `sink` until `setConsent(true)`, and nothing
/// tracked before opt-in is replayed afterwards.
public final class ConsentGatedAnalytics: Analytics {
  private let sink: any AnalyticsSink
  public private(set) var hasConsent = false

  public init(sink: any AnalyticsSink) {
    self.sink = sink
  }

  public func setConsent(_ granted: Bool) {
    hasConsent = granted
  }

  public func track(_ event: AnalyticsEvent) {
    guard hasConsent else { return }
    sink.send(event)
  }
}

extension AnalyticsEvent {
  /// `app_opened` from the taxonomy (packages/contracts/events/analytics/events.json);
  /// `consentRequired: true`, so the gate drops it until the user opts in.
  public static func appOpened(appVersion: String, coldStart: Bool) -> AnalyticsEvent {
    AnalyticsEvent(
      name: "app_opened",
      properties: [
        "platform": .string("ios"),
        "appVersion": .string(appVersion),
        "coldStart": .bool(coldStart),
      ]
    )
  }
}
